### PR TITLE
Encrypt integration - alternative approach

### PR DESCRIPTION
**Pull request by svendecabooter**
Alternative, simplified approach for integrating the field_encrypt module with the Encrypt module:
- removes the FieldEncryptMap logic altogether - we assume each field has properties that can encrypted by the Encrypt service. Perhaps this will need some extra validation or documentation, since it moves responsibility a bit more to the end user (but avoids the limitations of fixed field mappings)
- the field storage form now has 3 settings
  - whether to encrypt the field (checkbox)
  - which properties to encrypt (the "value" property in most cases - some reasonable defaults are provided)
  - the encryption profile to encrypt with
- upon encrypting fields, we check these settings, and encrypt the selected properties in case encryption was enabled, using the selected encryption profile.

This alternative approach makes the dependency on Encrypt module a bit less open-ended, but I don't think that's necessarily a bad thing...

### CODE REVIEW

**Comment by svendecabooter** [comment]
See also the listing of TODOs in https://github.com/d8-contrib-modules/field_encrypt/pull/19.
Perhaps it wasn't the best idea to create a new PR for the flow of the discussion, but it's easier to separate the 2 alternative approaches.

Some extra remarks, which were raised in https://github.com/d8-contrib-modules/field_encrypt/pull/19:
- global properties like ID and other internal properties don't seem to be returned when calling FieldStorageConfig::getPropertyDefinitions(), which the new code does. I only tested for a few common field types for now, so this needs some extra verification, but at first glance it seems we won't have to filter out those, since they aren't returned in the first place.
- base64 encoding seems needed to avoid database errors. The Key module provides base64 encoding / decoding of the keys, but the Encrypt module doesn't encode / decode the encrypted / decrypted strings.
  Perhaps we can add that as an optional setting in the EncryptService's encrypt() and decrypt() methods, instead of doing the base64 work in field_encrypt. But i'll leave it in for now, until we decided on that.

File: field_encrypt.module [hunk 1]
```diff
@@ -17,19 +27,49 @@ function field_encrypt_form_alter(&$form, \Drupal\Core\Form\FormStateInterface $
     $user = \Drupal::currentUser();
 
     if ($user->hasPermission('administer field encryption')) {
-      /**
-       * @var $field \Drupal\field\Entity\FieldStorageConfig
-       */
+       /* @var $field \Drupal\field\Entity\FieldStorageConfig */
       $field = $form_state->getFormObject()->getEntity();
+      $fieldSettings = \Drupal::service('field_encrypt.field_settings');
```
Comment (nerdstein) [inline field_encrypt.module]: Why is this a service? Shouldn't this just be pulled from CMI?

File: src/FieldEncryptFieldSettings.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  /**
+   * Get a list of reasonable default field properties eligible for encryption.
+   *
+   * @TODO: make this configurable via settings or hook?
+   *
+   * @return array
+   *   List of properties to select by default.
+   */
+  public function getDefaultProperties() {
```
Comment (nerdstein) [inline src/FieldEncryptFieldSettings.php]: I think this is what should be done in the settings form.

Upon install, the settings form should be populated with these things.

File: src/FieldEncryptFieldSettings.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace Drupal\field_encrypt;
+
+use Drupal\encrypt\EncryptionProfileManagerInterface;
+use Drupal\field\Entity\FieldStorageConfig;
+
+/**
+ * Service class to provide extra settings to field storage.
+ */
+class FieldEncryptFieldSettings {
```
Comment (nerdstein) [inline src/FieldEncryptFieldSettings.php]: I've reviewed this class and I honestly think we might be able to remove it. I like some of the helper functions but I think one of them will be rolled into the settings form, the others can likely be refactored...

Thoughts?

File: src/FieldEncryptProcessEntities.php [hunk 4]
```diff
@@ -121,33 +125,25 @@ protected function process_value($value = '', $service, $op = 'encrypt') {
...
    * @param string $op
-   * @param boolean $force if set, we don't check if encryption is enabled, we process the field anyway. This is used during batch processes.
+   *   The operation to perform (encrypt / decrypt).
+   * @param boolean $force
+   *   Whether to force the operation.
+   *   If set, we don't check if encryption is enabled, we process the field
+   *   anyway. This is used during batch processes.
    */
-  protected function process_field(\Drupal\Core\Field\FieldItemListInterface $field, $op = 'encrypt', $force = FALSE) {
+  protected function process_field(FieldItemListInterface $field, $op = 'encrypt', $force = FALSE) {
```
Comment (nerdstein) [inline src/FieldEncryptProcessEntities.php]: processField

File: src/FieldEncryptProcessEntities.php [hunk 5]
```diff
@@ -85,32 +87,34 @@ public function encrypt_entity(\Drupal\Core\Entity\ContentEntityInterface $entit
...
-   * @return string
+   *   The operation ("encrypt" or "decrypt")
    *
-   * TODO: If we can rely on the encrypt module providing an interface,
-   * we can include that here.
+   * @return string
+   *   The processed value.
    */
-  protected function process_value($value = '', $service, $op = 'encrypt') {
+  protected function process_value($value = '', EncryptionProfileInterface $encryption_profile, $op = 'encrypt') {
```
Comment (nerdstein) [inline src/FieldEncryptProcessEntities.php]: processValue

File: src/FieldEncryptProcessEntities.php [hunk 6]
```diff
@@ -76,7 +78,7 @@ protected function getFieldEncryptMap() {
    *
    * @param \Drupal\Core\Entity\ContentEntityInterface $entity
    */
-  public function encrypt_entity(\Drupal\Core\Entity\ContentEntityInterface $entity) {
+  public function encrypt_entity(ContentEntityInterface $entity) {
```
Comment (nerdstein) [inline src/FieldEncryptProcessEntities.php]: encryptEntity

File: src/FieldEncryptProcessEntities.php [hunk 7]
```diff
@@ -85,32 +87,34 @@ public function encrypt_entity(\Drupal\Core\Entity\ContentEntityInterface $entit
    *
    * @param \Drupal\Core\Entity\ContentEntityInterface $entity
    */
-  public function decrypt_entity(\Drupal\Core\Entity\ContentEntityInterface $entity) {
+  public function decrypt_entity(ContentEntityInterface $entity) {
```
Comment (nerdstein) [inline src/FieldEncryptProcessEntities.php]: decryptEntity

File: src/FieldEncryptProcessEntities.php [hunk 8]
```diff
@@ -173,21 +169,23 @@ protected function process_field(\Drupal\Core\Field\FieldItemListInterface $fiel
...
-      $map = $this->fieldEncryptMap[$field_type];
-      foreach($map as $value_name => $service) {
-        if(isset($value[$value_name])){
-          $value[$value_name] = $this->process_value($value[$value_name], $service, $op);
+    $encryption_profile_id = $storage->getThirdPartySetting('field_encrypt', 'encryption_profile', []);
+    $encryption_profile = $this->encryptionProfileManager->getEncryptionProfile($encryption_profile_id);
+
+    // Process the field with the given encryption provider.
+    foreach ($field_value as &$value) {
+      $properties = $storage->getThirdPartySetting('field_encrypt', 'properties', []);
```
Comment (nerdstein) [inline src/FieldEncryptProcessEntities.php:176]: Does this filter out properties that have been set to encrypt? I am not seeing or understanding how this block of code is doing so.

File: src/FieldEncryptProcessEntities.php [hunk 9]
```diff
@@ -197,9 +195,11 @@ protected function process_field(\Drupal\Core\Field\FieldItemListInterface $fiel
    * Both processes are very similar, so we bundle the field processing part.
    *
    * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The entity to process.
    * @param string $op
+   *   The operation to perform (encrypt / decrypt).
    */
-  protected function process_entity(\Drupal\Core\Entity\ContentEntityInterface $entity, $op = 'encrypt') {
+  protected function process_entity(ContentEntityInterface $entity, $op = 'encrypt') {
```
Comment (nerdstein) [inline src/FieldEncryptProcessEntities.php]: processEntity

File: src/FieldEncryptProcessEntities.php [hunk 10]
```diff
@@ -216,7 +216,9 @@ protected function process_entity(\Drupal\Core\Entity\ContentEntityInterface $en
    * This is performed when field storage settings are updated.
    *
    * @param $entity_type
+   *   The entity type.
    * @param $field_name
+   *   The name of the field to encrypt.
    */
   public function encrypt_stored_field($entity_type, $field_name) {
```
Comment (nerdstein) [inline src/FieldEncryptProcessEntities.php]: encryptStoredField

File: src/FieldEncryptProcessEntities.php [hunk 11]
```diff
@@ -239,8 +243,11 @@ public function decrypt_stored_field($entity_type, $field_name) {
...
    *
    * @param $entity_type
+   *   The entity type.
    * @param $field_name
-   * @param string $op (encrypt / decrypt)
+   *   The name of the field to update.
+   * @param string $op
+   *   The operation to perform (encrypt / decrypt).
    */
   protected function update_stored_field($entity_type, $field_name, $op = 'encrypt') {
```
Comment (nerdstein) [inline src/FieldEncryptProcessEntities.php]: updateStoredField

**Comment by nerdstein** [comment]
Finished reviewing and left comments.

YES this is much closer to what I was thinking. I really like it.

File: src/FieldEncryptProcessEntities.php (hunk 8 above)
Comment (svendecabooter) [inline src/FieldEncryptProcessEntities.php:176]: It selects the properties for a given field that should be encrypted (e.g. text, summary) and encrypts them, while leaving properties that have not been set up for encryption (e.g. text_format) remain unchanged.

**Comment by svendecabooter** [comment]
Added settings form to select default properties per field type, to be used when setting up any field for field encryption. Upon installing the module, sensible defaults are already set for core Drupal field types.

**Comment by nerdstein** [comment]
This is perfect, I'm merging this in to get this going. I love it

**Pull request merged**